File: src/principlesGate.ts
import type { BlacklistEntry } from './rules.js';

export interface GateResult {
  allowed: boolean;
  reason?: string;
}

/**
 * Deterministic blacklist check. Returns denied with a reason if the
 * (transport, action) pair appears in the blacklist.
 */
export function checkBlacklist(
  blacklist: BlacklistEntry[],
  transport: string,
  action: string,
): GateResult {
  const t = transport.toLowerCase();
  const a = action.toLowerCase();
  const match = blacklist.find(
    (e) => e.transport.toLowerCase() === t && e.action.toLowerCase() === a,
  );
  if (match) {
    return { allowed: false, reason: `Blacklisted: (${match.transport}, ${match.action})` };
  }
  return { allowed: true };
}

File: python/steward/principles_gate.py
"""Deterministic blacklist gate enforced before dispatching any approved action."""
from __future__ import annotations

from dataclasses import dataclass

from steward.rules import BlacklistEntry


@dataclass
class GateResult:
    allowed: bool
    reason: str | None = None


def check_blacklist(
    blacklist: list[BlacklistEntry],
    transport: str,
    action: str,
) -> GateResult:
    t = transport.lower()
    a = action.lower()
    for entry in blacklist:
        if entry["transport"].lower() == t and entry["action"].lower() == a:
            return GateResult(
                allowed=False,
                reason=f"Blacklisted: ({entry['transport']}, {entry['action']})",
            )
    return GateResult(allowed=True)
